metaflow/plugins/aws/batch/batch_decorator.py

```diff
@@ -421,6 +421,86 @@ def _wait_for_mapper_tasks(self, flow, step_name):
         TIMEOUT = 600
         last_completion_timeout = time.time() + TIMEOUT
         print("Waiting for batch secondary tasks to finish")
+
+        # Prefer Batch API when metadata is local (nodes can't share local metadata files).
+        # If metadata isn't bound yet but we are on Batch, also prefer Batch API.
+        md = getattr(self, "metadata", None)
+        if md is not None and md.TYPE == "local":
+            return self._wait_for_mapper_tasks_datastore(
+                flow, step_name, last_completion_timeout
+            )
+        if md is None and "AWS_BATCH_JOB_ID" in os.environ:
+            return self._wait_for_mapper_tasks_datastore(
+                flow, step_name, last_completion_timeout
+            )
+        return self._wait_for_mapper_tasks_metadata(
+            flow, step_name, last_completion_timeout
+        )
+
+    def _wait_for_mapper_tasks_datastore(
+        self, flow, step_name, last_completion_timeout
+    ):
+        """
+        Poll the shared datastore (S3) for DONE markers for each mapper task.
+        This avoids relying on a metadata service or local metadata files.
+        """
+        from metaflow.datastore.task_datastore import TaskDataStore
+
+        pathspecs = getattr(flow, "_control_mapper_tasks", [])
+        total = len(pathspecs)
+        if total == 0:
+            print("No mapper tasks discovered for datastore wait; returning")
+            return True
+
+        print("Waiting for mapper DONE markers in datastore for %d tasks" % total)
+        poll_sleep = 3.0
+        while last_completion_timeout > time.time():
+            time.sleep(poll_sleep)
+            completed = 0
+            for ps in pathspecs:
+                try:
+                    parts = ps.split("/")
+                    if len(parts) == 3:
+                        run_id, step, task_id = parts
+                    else:
+                        # Fallback in case of unexpected format
+                        run_id, step, task_id = self.run_id, step_name, parts[-1]
+                    tds = TaskDataStore(
+                        self.flow_datastore,
+                        run_id,
+                        step,
+                        task_id,
+                        mode="r",
+                        allow_not_done=True,
+                    )
+                    if tds.has_metadata(TaskDataStore.METADATA_DONE_SUFFIX):
+                        completed += 1
+                except Exception as e:
+                    self.logger.warning("Datastore wait: error checking %s: %s", ps, e)
```

Comment:
`self.logger` doesn't actually have any methods, it is just `click.secho` being passed in. This also adds unnecessary (duplicate) timestamps to the log lines so sticking to `print` for now is fine.